### ISSUE TITLE
v0.11.3 is broken ?

### ISSUE THREAD

**Issue by xfournet** [ZJONSSON/node-unzipper#306]
With previous version (v0.11.2) the following code was working

```js
import { Open as ZipOpen } from 'unzipper';

const unzip = await ZipOpen.file('dummy.zip');
const files = unzip.files;
files.forEach((entry) => {
  console.log(entry.name);
});
```
-> print one line per file in the zip file, with the corresponding zip file entry name

With v0.11.3 this fail:
```
file:///D:/sources/test/testUnzipper.js:5
files.forEach((entry) => {
      ^

TypeError: files.forEach is not a function
    at file:///D:/sources/test/testUnzipper.js:5:7
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v21.7.3
```

First, there is a undocumented breaking because `unzip.files` is now a promise, so I need to update the code like this
```js
import { Open as ZipOpen } from 'unzipper';

const unzip = await ZipOpen.file('dummy.zip');
const files = await unzip.files;
files.forEach((entry) => {
  console.log(entry.name);
});
```

But this still fails:
```
file:///D:/sources/test/testUnzipper.js:6
  console.log(entry.path);
                    ^

TypeError: Cannot read properties of undefined (reading 'path')
    at file:///D:/sources/test/testUnzipper.js:6:21
    at Array.forEach (<anonymous>)
    at file:///D:/sources/test/testUnzipper.js:5:7
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v21.7.3
```

this is because all elements in the `files` array is undefined

**Comment by OmarShehata** [comment]
Thanks for opening this @xfournet , I also was confused why the official example seemed to be broken. I downgraded to `"unzipper": "^0.10.11"` which seems to work.

My use case was unzipping and extracting a couple specific files to disk. This is what ended up working for me:

```javascript
import unzipper from 'unzipper'
import * as fs from 'fs';
import path from 'path'

async function run() {
    // Extract the list of files from the zip
    const directory = await unzipper.Open.file('./archive.zip');
    // Get a specific file
    const file = directory.files.find(item => item.path.includes('README.txt'))
    // Extract the file in the same directory & write it to the same relative location as inside the zip
    // file.path for example may be `data/README.txt`
    await streamFileToDisk(file, `./${file.path}`)
}
run()


/// Helper
function streamFileToDisk(file, outputPath) {
    return new Promise((resolve, reject) => {
        // Extract directory path from outputPath
        const directory = path.dirname(outputPath);
        // Ensure directory exists or create it
        fs.mkdir(directory, { recursive: true }, (err) => {
            if (err) {
                return reject(err);
            }

            const fileStream = file.stream();  
            const outputStream = fs.createWriteStream(outputPath);

            fileStream.pipe(outputStream)
                .on('error', (error) => {
                    console.error('Error while streaming file:', error);
                    reject(error);
                })
                .on('finish', () => {
                    resolve();
                });
        });
    });
}
```

**Comment by ZJONSSON** [comment]
Thanks @xfournet .  The property you are looking for is not `.name` but rather `.path`.   I looked at all the examples in the readme and they all show `.path`.   Also the typescript types confirm this has been the case for a long time (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/unzipper/index.d.ts#L99)

Can you investigate why you are seeing files as a promise (what you say is breaking change).   Before returning the directory object all the properties are awaited on, so there should be no variables remaining as promises due to https://github.com/ZJONSSON/node-unzipper/blob/master/lib/Open/directory.js#L228

You example works like this for me (removed files = await unzip.files, and changed `.name` to `.path`):
```js
  const unzip = await ZipOpen.file('./testData/compressed-directory-entry/archive.zip');
  unzip.files.forEach((entry) => {
    console.log(entry.path);
  });
```

**Comment by ZJONSSON** [comment]
@OmarShehata, can you point out which specific examples were failing for you?

**Comment by ZJONSSON** [comment]
Please reopen the ticket if you have specific examples, and we can look at them. The test directory is also a good source for examples.

**Comment by xfournet** [comment]
Hi @ZJONSSON ,

yes sorry there was in mistake in my copy path/paste to extract the minimal repro case, indeed it's `path` that is used in my code, not `name`.

Here is the fixed repro case : https://codesandbox.io/p/devbox/quirky-chaum-zrhprs to be sure we can observe the same thing

As you can see there is a first error : 
![image](https://github.com/ZJONSSON/node-unzipper/assets/461943/727f1909-fb31-4e7e-bc50-75973a0b10fe)

`files.forEach is not a function` -> you can easily see that files doesn't return an array as expected from the TS definition but instead it's a promise that is returned -> that's why it's seems there is a regression (if it's unexpected) or a breaking change (if it was expected).

even if i resolve the promise, i keep blocked, since the resolved promise return an array filled with undefined entries.

![image](https://github.com/ZJONSSON/node-unzipper/assets/461943/9fc5c5c4-befc-485e-affa-6ca49f61ee14)

could you reopen this issue please ? this is basically the same exemple that i initially provided, i cannot be more specific and it's clearly show the regression IMHO

from the code you pointed, i see it use Promise.mapSeries (https://github.com/ZJONSSON/node-unzipper/blob/master/lib/Open/directory.js#L179) that is unknown from Node.js. It seems to be from bluebird, that doesn't seems to be used in unzipper ?
It's very confusing since it's not the code that has been release in 0.11.3, as you can see here : 

![image](https://github.com/ZJONSSON/node-unzipper/assets/461943/a664516d-db29-4da1-a60f-1bd982726cbe)

* it's Promise.all that is used rather than Promise.mapSeries in the master branch
* the code that resolved the promise has been commented ?? `return vars //romise.props(vars);`

it look strange, could you clarify if the release and/or the master branch are ok ?

**Comment by ZJONSSON** [comment]
Wow thank you - I was thrown off the right path by the .path vs .name.   You are indeed correct, there was a divergence between master and published version (apologies).   I will be moving the publish process to github actions soon to prevent this.  In the meantime I have now released 0.11.4 which should be exact match of master.

Thanks again for the diligent investigation.  Please let me know if there are any issues with the new release

**Comment by xfournet** [comment]
Thanks for quickly taking care of it, v0.11.4 is ok.
